### PR TITLE
Add validator to the mutation compactor

### PR DESCRIPTION
**Pull request by denesb**
Fragment reordering and fragment dropping bugs have been plaguing us since forever. To fight them we added a validator to the sstable write path to prevent really messed up sstables from being written.
This series adds validation to the mutation compactor. This will cover reads and compaction among others, hopefully ridding us of such bugs on the read path too.
This series fixes some benign looking issues found by unit tests after the validator was added -- although how benign a producer emitting two partition-ends depends entirely on how the consumer reacts to it, so no such bug is actually benign.

Fixes: https://github.com/scylladb/scylladb/issues/11174

### CODE REVIEW

**Comment by denesb** [comment]
I want to start a BYO job with this PR, don't merge until we see the result of this run. Last time I added validation to the read path CI/promotion didn't catch anything, but full dtest run failed.

**Comment by denesb** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2212/

CI found further invalid mutation fragment streams. I will look into them.

**Comment by denesb** [comment]
Found https://github.com/scylladb/scylladb/issues/11561.

**Comment by denesb** [comment]
New in v2:
* Fix the newly found https://github.com/scylladb/scylladb/issues/11561 (thanks to the validator).

**Comment by denesb** [comment]
Its the flaky CDC test [Sanity Tests / cdc_test.TestCdc.test_cluster_expansion_with_cdc[Multi_DC_cluster]](https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2280/testReport/junit/cdc_test/TestCdc/Sanity_Tests___test_cluster_expansion_with_cdc_Multi_DC_cluster_/) again.

**Comment by denesb** [comment]
Rebased on master.

File: readers/mutation_reader.cc
Comment (xemul) [inline readers/mutation_reader.cc:192]: nit: There's `logger::is_enabled(level)` just for this

**Comment by denesb** [comment]
New in v3:
* Use `logger::is_enabled()` instead of open-coding the same.

**Comment by denesb** [comment]
Started BYO full dtest run: https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/1511/
Don't merge until this is finished.

**Comment by denesb** [comment]
BYO is currently broken and I also just realized we have a list of tests that are known to be fail, so I will have to go over each failed test and check whether it failed due to the validator or not.
All this is to say that green-lighting the validator itself will take ages, so I will start extracting the individual fixes in the meanwhile, so they can get in.

**Comment by denesb** [comment]
Split out:
* https://github.com/scylladb/scylladb/pull/11603
* https://github.com/scylladb/scylladb/pull/11604

**Comment by denesb** [comment]
And https://github.com/scylladb/scylladb/pull/11614.

**Comment by denesb** [comment]
Rebased on master, on top of the now merged split out PRs. Will kick-off BYO once CI approves.

**Comment by denesb** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2526/

Resharding test timed out. Re-kicked CI.

File: mutation_compactor.hh
Comment (tgrabiec) [inline mutation_compactor.hh:306]: We should add `position_in_partition_view::for_partition_start()`.

**Review by tgrabiec — APPROVED** [review]

File: mutation_compactor.hh
Comment (denesb) [reply mutation_compactor.hh:306]: Yes.

**Comment by denesb** [comment]
This was dequeued as it was suspected that it causes failures in next promotion.

**Comment by denesb** [comment]
v4:
* Add (and use) position_in_partition{_view}::for_partition_{start,end}().
* Raise validation level to token.

The raised validation level found further violations, possibly due to https://github.com/scylladb/scylladb/pull/11886, so don't merge yet.

**Comment by denesb** [comment]
v5:
* Fix token order violations found in tests. All are benign:
    - scrub/validation tests were triggering validation errors as these tests work with corrupt data
    - mutation_query_test::partition_limit_test didn't sort its input data
* Disable validation on the compaction path -- we already have a validator in the sstable writer.

**Comment by denesb** [comment]
@scylladb/scylla-maint ping.

File: mutation_compactor.hh
Comment (avikivity) [inline mutation_compactor.hh:301]: This constructs a string in mutation_fragment_stream_validating_filter, adding an allocation to all reads. Please follow up with changing that to a string_view (and verifying all callers indeed pass a constant; you can append the "sv" literal operator as a marker).

Comment (denesb) [reply mutation_compactor.hh:301]: Some callers pass in string created with `format()`. We can do the usual trick of having a string and string_view in the class to avoid allocating when possible.

Comment (denesb) [reply mutation_compactor.hh:301]: Followup PR is here: https://github.com/scylladb/scylladb/pull/12042

**Pull request merged**
